File: algorithms/cpp/removeComments/removeComments.cpp
// Source: https://leetcode.com/problems/remove-comments/
// Author: Miao Zhang
// Date:   2021-03-04

class Solution {
public:
    vector<string> removeComments(vector<string>& source) {
        vector<string> res;
        bool blocked = false;
        string out = "";
        for (string line: source) {
            for (int i = 0; i < line.size(); i++) {
                if (!blocked) {
                    if (i == line.size() - 1) {
                        out += line[i];
                    } else {
                        string t = line.substr(i, 2);
                        if (t == "/*") {
                            blocked = true;
                            i++;
                        } else if (t == "//") {
                            break;
                        } else {
                            out += line[i];
                        }
                    }
                } else {
                    if (i < line.size() - 1) {
                        string t = line.substr(i, 2);
                        if (t == "*/") {
                            blocked = false;
                            i++;
                        }
                    }
                }
            }
            if (!out.empty() && !blocked) {
                res.push_back(out);
                out = "";
            }
        }
        return res;
    }
};

File: algorithms/python/removeComments/removeComments.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/remove-comments/
# Author: Miao Zhang
# Date:   2021-03-04

class Solution:
    def removeComments(self, source: List[str]) -> List[str]:
        blocked = False
        res = []
        for line in source:
            i = 0
            if not blocked:
                newline = []
            while i < len(line):
                if line[i: i+2] == '/*' and not blocked:
                    blocked = True
                    i += 1
                elif line[i: i+2] == '*/' and blocked:
                    blocked = False
                    i += 1
                elif not blocked and line[i: i+2] == '//':
                    break
                elif not blocked:
                    newline.append(line[i])
                i += 1
            if newline and not blocked:
                res.append("".join(newline))
        return res
